core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,747 @@
+    """Predict classes for each program in dataset using KNN.
+
+    Args:training_data
+        training_data: dict. A dictionary containing training data.
+            Structure of training_data is as follows,
+                {
+                    ID: {
+                        'source': str. Source code of program.
+                        'class': int. Associated feedback class with program.
+                        'tokens': list(str). Token list of program.
+                        'k_gram_hashes': list(int). K-gram hash values of
+                            program.
+                        'fingerprint': list(int). Extracted fingerprint of
+                            program.
+                        'nearest_neighbours': list(int). The indices of the
+                            programs that are most similar to this one
+                    } for each program in training_data.
+                }
+        top_neighbours: int. No. of nearest neighbours to consider for KNN.
+
+    Returns:
+        float. Average no. of times a class has to appear in nearest neighbour
+            for it to be correct prediction.
+        list(int). IDs of training_data programs which are misclassified by KNN.
+    """
+    training_data = add_k_nearest_neighbours(training_data, top_neighbours)
+
+    # No. of times a class has to appear in nearest neighbours of a program
+    # so that prediction is correct.
+    occurrence = 0
+
+    # Keep record of misclassified programs' ID.
+    misclassified_points = []
+
+    for pid in training_data:
+        nearest_neighbours = training_data[pid]['nearest_neighbours']
+        if pid in nearest_neighbours:
+            nearest_neighbours.remove(pid)
+        nearest_classes = [
+            training_data[i]['class'] for (i, _) in nearest_neighbours]
+        cnt = collections.Counter(nearest_classes)
+        common = cnt.most_common(1)
+        training_data[pid]['prediction'] = common[0][0]
+
+        if training_data[pid]['prediction'] == training_data[pid]['class']:
+            # If the prediction is correct, then increment 'occurrence' by the
+            # number of times the correct class has to appear in nearest
+            # neighbours. During prediction this value is used as threshold
+            # and prediction class must appear more than occurrence times
+            # in nearest neighbors to give assurance of correct classification.
+            occurrence += common[0][1]
+        else:
+            # Else mark the ID of program so that it will be
+            # trained using SVM later.
+            misclassified_points.append(pid)
+
+    # Calculate average of occurrence. This is used during prediction so
+    # that if winner of nearest neighbours appears less than the occurrence
+    # we consider that KNN has failed in prediction.
+    occurrence /= float(len(training_data))
+
+    return occurrence, misclassified_points
+
+
+class CodeClassifier(base.BaseClassifier):
```

Comment:
Where is the predict() method? We need some mechanism of testing the complete classifier on backend, how will we do that?